### ISSUE TITLE
Prange requires return value.

### ISSUE THREAD

**Issue by stuartarchibald** [numba/numba#2474]
Reported on the mailing list:
https://groups.google.com/a/continuum.io/forum/#!topic/numba-users/brw0TVPejdo

Ping @ehsantn @ninegua .

This fails:
```python
import numpy as np
import numba

N = 1000
x = np.random.rand(N)

@numba.njit(parallel=True)
def add_one( x, N ):
    for i in numba.prange(N):
        x[i] = x[i] + 1
    return
#   return x : fixes the compilation                                           

add_one( x, N )
```
with:
```python
Traceback (most recent call last):
  File "<install_path>/numba/errors.py", line 243, in new_error_context
    yield
  File "<install_path>/numba/lowering.py", line 216, in lower_block
    self.lower_inst(inst)
  File "<install_path>/numba/lowering.py", line 360, in lower_inst
    parfor.lower_parfor_parallel(self, inst)
  File "<install_path>/numba/npyufunc/parfor.py", line 49, in _lower_parfor_parallel
    assert parfor.params
AssertionError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "ml10.py", line 14, in <module>
    add_one( x, N )
  File "<install_path>/numba/dispatcher.py", line 307, in _compile_for_args
    return self.compile(tuple(argtypes))
  File "<install_path>/numba/dispatcher.py", line 579, in compile
    cres = self._compiler.compile(args, return_type)
  File "<install_path>/numba/dispatcher.py", line 80, in compile
    flags=flags, locals=self.locals)
  File "<install_path>/numba/compiler.py", line 740, in compile_extra
    return pipeline.compile_extra(func)
  File "<install_path>/numba/compiler.py", line 360, in compile_extra
    return self._compile_bytecode()
  File "<install_path>/numba/compiler.py", line 699, in _compile_bytecode
    return self._compile_core()
  File "<install_path>/numba/compiler.py", line 686, in _compile_core
    res = pm.run(self.status)
  File "<install_path>/numba/compiler.py", line 246, in run
    raise patched_exception
  File "<install_path>/numba/compiler.py", line 238, in run
    stage()
  File "<install_path>/numba/compiler.py", line 621, in stage_nopython_backend
    self._backend(lowerfn, objectmode=False)
  File "<install_path>/numba/compiler.py", line 576, in _backend
    lowered = lowerfn()
  File "<install_path>/numba/compiler.py", line 563, in backend_nopython_mode
    self.flags)
  File "<install_path>/numba/compiler.py", line 858, in native_lowering_stage
    lower.lower()
  File "<install_path>/numba/lowering.py", line 135, in lower
    self.lower_normal_function(self.fndesc)
  File "<install_path>/numba/lowering.py", line 176, in lower_normal_function
    entry_block_tail = self.lower_function_body()
  File "<install_path>/numba/lowering.py", line 201, in lower_function_body
    self.lower_block(block)
  File "<install_path>/numba/lowering.py", line 216, in lower_block
    self.lower_inst(inst)
  File "<python_path>/lib/python3.6/contextlib.py", line 100, in __exit__
    self.gen.throw(type, value, traceback)
  File "<install_path>/numba/errors.py", line 249, in new_error_context
    six.reraise(type(newerr), newerr, sys.exc_info()[2])
  File "<install_path>/numba/six.py", line 658, in reraise
    raise value.with_traceback(tb)
  File "<install_path>/numba/errors.py", line 243, in new_error_context
    yield
  File "<install_path>/numba/lowering.py", line 216, in lower_block
    self.lower_inst(inst)
  File "<install_path>/numba/lowering.py", line 360, in lower_inst
    parfor.lower_parfor_parallel(self, inst)
  File "<install_path>/numba/npyufunc/parfor.py", line 49, in _lower_parfor_parallel
    assert parfor.params
numba.errors.LoweringError: Failed at nopython (nopython mode backend)

File "ml10.py", line 9
[1] During: lowering "[LoopNest(index_variable=parfor_index.3, range=0,N,1 correlation=-1)]{14: <ir.Block at ml10.py (9)>}Var(parfor_index.3, ml10.py (9))" at ml10.py (9)
```

**Comment by ehsantn** [comment]
This is already fixed in the [array-comprehension PR](https://github.com/numba/numba/pull/2400) @stuartarchibald .

**Comment by stuartarchibald** [comment]
@ehsantn great, thanks. I don't recall seeing a test case for it ? Did I miss it?

**Comment by ehsantn** [comment]
The issue was that alias analysis couldn't properly figure out the computations affecting mutable input argument `x`, so dead code elimination would think the whole loop body is dead. Having return statement makes it explicit that `x` computations are alive. The array comprehension test cases cover this issue.

**Comment by stuartarchibald** [comment]
Ah I see, thanks, that makes sense. Will close when #2400 is merged.

**Comment by stuartarchibald** [comment]
Fixed in #2400 closing.
